docs/architecture.md

```diff
@@ -23,67 +23,87 @@
 
-### Protocol
+### Connection
 
-Protocols define how messages are represented and encoded for transport. They also define the rules to which messages have to adhere in a message sequence. For instance, a protocol might have a message of type START and FINISH. Then the rules could prescribe that a message of type FINISH must be preceded by a message of type START.
+A connection wraps an external SDK or API and manages the messaging. It allows the agent to connect to an external service which has a Python SDK or API. 
+
+### Skill
 
+Skills deliver economic value to the AEA by allowing an agent to encapsulate and call any kind of code. It encapsulates Handlers, Behaviours, and Tasks.
 
-### Mailbox
+* Handler: each skill has a single Handler which is responsible for the registered protocol messaging. By understanding the requirements contained in Envelopes, the Handler reacts appropriately to message requests.
+* Behaviour: one or more Behaviours encapsulate sequences of actions that cause interactions with other agents initiated by the framework. 
+* Task: one or more Tasks encapsulate background work internal to the agent.
 
-TBC.
+### MailBox
+
+A MailBox contains InBox and OutBox queues which manage Envelopes.
 
 ### Envelope
 
-The `Envelope` is the core object which agents use to communicate with each other. An `Envelope` has four attributes:
+An Envelope is the core object which agents use to communicate with each other. It has four attribute parameters:
+
+* `to`: defines the destination address.
 
-* `to`: defines the destination address
+* `sender`: defines the sender address.
 
-* `sender`: defines the sender address
+* `protocol_id`: defines the id of the protocol.
 
-* `protocol_id`: defines the protocol_id
+* `message`: is a bytes field which holds the message in serialized form.
 
-* `message`: is a `bytes` field to hold the message in serialized form.
+The Envelope encodes and decodes messages.
 
 
-### MainLoop
+## Agent 
 
-TBC.
+### Main loop
 
-### ProtocolMessage
+The `_run_main_loop()` function in the `Agent` class performs a series of activities while the `Agent` state is not stopped.
 
-TBC. 
+* `act()`: this function calls the `act()` function of all registered Behaviours.
+* `react()`: this function grabs all Envelopes waiting in the InBox queue and calls the `handle()` function on them.
+* `update()`: this function loops through all the Tasks and executes them.
 
-### Director
+## Resources 
 
-TBC. 
+The `Resources` class implements all resources for an agent. These come in the form of registries.
 
-### Orchestrator
+The specific classes are in the `registries/base.py` module.
 
-TBC.
+* ProtocolRegistry.
+* HandlerRegistry. 
+* BehaviourRegistry.
+* TaskRegistry.
+
+
+## Director
+
+!!! TODO 
+
+## Orchestrator
```

Comment:
We call this `Filter` now as it will have a passive routing role (it makes no choices, just implements the message routing to the correct helpers)